apis/cni/v1alpha1/antreaconfig_webhook.go

```diff
@@ -0,0 +1,135 @@
+)
+
+// log is for logging in this package.
+var antreaconfiglog = logf.Log.WithName("antreaconfig-resource")
+
+func (r *AntreaConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+var _ webhook.Defaulter = &AntreaConfig{}
+var _ webhook.Validator = &AntreaConfig{}
+
+// +kubebuilder:webhook:path=/mutate-cni-tanzu-vmware-com-v1alpha1-antreaconfig,mutating=true,failurePolicy=fail,groups=cni.tanzu.vmware.com,resources=antreaconfigs,verbs=create;update,versions=v1alpha1,name=mantreaconfig.kb.io,admissionReviewVersions=v1,sideEffects=None
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *AntreaConfig) Default() {
+	antreaconfiglog.Info("default", "name", r.Name)
+	// No-op for default
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cni-tanzu-vmware-com-v1alpha1-antreaconfig,mutating=false,failurePolicy=fail,groups=cni.tanzu.vmware.com,resources=antreaconfigs,versions=v1alpha1,name=vantreaconfig.kb.io,admissionReviewVersions=v1,sideEffects=None
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AntreaConfig) ValidateCreate() error {
+	antreaconfiglog.Info("validate create", "name", r.Name)
+
+	var allErrs field.ErrorList
+
+	if !r.Spec.Antrea.AntreaConfigDataValue.FeatureGates.AntreaProxy &&
+		r.Spec.Antrea.AntreaConfigDataValue.FeatureGates.EndpointSlice {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "antrea", "config", "featureGates", "EndpointSlice"),
+				r.Spec.Antrea.AntreaConfigDataValue.FeatureGates.EndpointSlice,
+				"field cannot be enabled if AntreaProxy is disabled"),
+		)
+	}
+
+	if r.Spec.Antrea.AntreaConfigDataValue.NoSNAT &&
+		(r.Spec.Antrea.AntreaConfigDataValue.TrafficEncapMode == "encap" ||
+			r.Spec.Antrea.AntreaConfigDataValue.TrafficEncapMode == "hybrid") {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "antrea", "config", "noSNAT"),
+				r.Spec.Antrea.AntreaConfigDataValue.NoSNAT,
+				"field must be disabled for encap and hybrid mode"),
+		)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "cni.tanzu.vmware.com", Kind: "AntreaConfig"}, r.Name, allErrs)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *AntreaConfig) ValidateUpdate(old runtime.Object) error {
+	antreaconfiglog.Info("validate update", "name", r.Name)
+
+	oldObj, ok := old.(*AntreaConfig)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("Expected an AntreaConfig but got a %T", oldObj))
+	}
+
+	var allErrs field.ErrorList
+
+	// Check for changes to immutable fields and return errors
+	if !reflect.DeepEqual(r.Spec.Antrea.AntreaConfigDataValue.DefaultMTU,
+		oldObj.Spec.Antrea.AntreaConfigDataValue.DefaultMTU) {
```

Comment:
@lubronzhan @jayunit100 - Do you know if antrea defaultMTU can be changed after cluster create?

Reply:
@yzaccc can you make the final call on this one ? thanks

Reply:
Zac is on PTO, but MTU can be modified in antrea.  In the case that you set it, and delete an `antrea-agent`, the new agent will have a new MTU. 

Im thus going to say it should allow mutation.  In general i think all params should allow mutation unless we specify otherwise - -

Reply:
Thanks for the info. will remove the mutation check as it is allowed

Reply:
@jayunit100 , so are there any immutable fields for antrea and calico?

Reply:
Looks like this is already merged. But yea, MTU is mutable, should not be checked here.